src/data/nat/basic.lean

```diff
@@ -1674,4 +1674,56 @@ lemma decreasing_induction_succ_left {P : ℕ → Sort*} (h : ∀n, P (n+1) →
 by { rw [subsingleton.elim mn (le_trans (le_succ m) smn), decreasing_induction_trans,
          decreasing_induction_succ'] }
 
+/-- `log hb n`, is the logarithm of natural number
+`n` in base `b`. If `n` is not a power of `b`, it returns the
+largest `k:ℕ` such that `b^k ≤ n`. -/
+def log (b : ℕ) : ℕ → ℕ
+| n :=
+  if h : b ≤ n ∧ 1 < b then
+    have n / b < n,
+      from div_lt_self
+        (nat.lt_of_lt_of_le (lt_trans zero_lt_one h.2) h.1) h.2,
+    log (n / b) + 1
+  else 0
```

Comment:
I wonder if we could get @shingtaklam1324's help integrating this with `data.nat.digits`? Ideally `log` would just be defined as the length of the digits.

Reply:
I don't see the benefit of doing that. Can you expand?

Reply:
I think the behaviour would be different between the two implementations. Would this matter?

If `b = 0`, then using the length of digits would return `1`, but this returns `0`. If `b = 1`, this returns `0` as well, but length of digits is `n`. Outside of these edge cases, it'd be the same.

I suppose some of these lemmas are similar to what's already been proven in `data.nat.digits`? Since some of the existing lemmas there refer to the length of digits. Although it's similar enough that I think shouldn't be too difficult to prove these from what we have in `data.nat.digits`

Reply:
I think `exp_log_le_self` is the same as `base_pow_length_digits_le` (except the power being off by 1). `exp_le_iff_le_log` can come from `log_exp` and existing lemmas in `data.nat.digits`. `exp_succ_log_gt_self` is the same as `lt_base_pow_length_digits`.

Reply:
But even then, I'm not sure if it makes sense to refactor these lemmas, rather it might make more sense to refactor `data.nat.digits`. These lemmas would have to be moved to `data.nat.digits` (or a file that imports it), and whatever files that uses these will need to import all of digits and it's imports. I don't know if this will cause any issues.

Reply:
But equally, it might make more sense to refactor `data.nat.digits` to add lemmas that use `log`, and prove that `(digits b m).length = log b m + 1`.

Reply:
@shingtaklam1324 and @semorrison, I don't think it's necessary at this point to involve `digits` in `random` because of how superficial my use of `log` is. I think I will involve `digits` when I need to prove something more significant about the number of bits I generate in `bitvec`

Reply:
@DanielFabian fixed. I think you meant "the *declaration* and the descriptions don't agree"

Reply:
@cipher1024 , no problem. That was pretty much what I was suggesting to do in my last comment.

Reply:
My complaint is more that we're duplicating code unnecessarily.

Reply:
I don't feel like this is duplication. I think it's worth connecting the two but `log` is a simpler function to reason about. I feel like we have a better separation of concern if we keep the two definitions.

Reply:
Ok!